core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -165,13 +171,13 @@ public boolean add(T object) {
     // Locking this object protects against removing the exact object, but does not protect against
     // removing a distinct, but equivalent object.
     synchronized (object) {
-      if (!mObjects.addIfAbsent(object)) {
+      if (mPrimaryIndex.containsObject(object)) {
```

Comment:
This is different behavior now. Before, the addIfAbsent would allow the check and the add to be done atomically. Now, it is checked here, and added later. Is that ok?

Also, in the previous behavior, the object to add was always added to `mObjects` first, before any indexes. Now, there is no deterministic order of which index gets the object first. How does that affect correctness?